lib/min_heap.rb

```diff
@@ -44,28 +49,50 @@ def to_s
   end
 
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: 1
+  # Space complexity: 1
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    return @store.empty?
   end
 
   private
 
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: log n
+  # Space complexity: log n
   def heap_up(index)
-    
+    return nil if index == 0
+    parent = (index - 1) / 2
+    if @store[index].key < @store[parent].key
+      swap(index, parent)
+      heap_up(parent)
+    end
   end
 
   # This helper method takes an index and 
   #  moves it up the heap if it's smaller
   #  than it's parent node.
+  # Time complexity: log n
+  # Space complexity: log n
   def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+    left = (index * 2) + 1
+    right = (index * 2) + 2
+
+    return @store if @store[left].nil?
+    return nil if left >= @store.length
+    
+    if @store[left].key < @store[index].key
+      # recursive case - left node
+      swap(left, index)
+      heap_down(left)
+    end
+    if right < @store.length && @store[right].key < @store[index].key
+      # recursive case - right node
+      swap(right, index)
+      heap_down(right)
+    end
```

Comment:
This works, but you might engage in extra swaps if the sequential if statements.